### PR TITLE
[Workspace]Add workspace id in basePath

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/workspace/public/workspace_client.ts [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+  /**
+   * Search for workspaces
+   *
+   * @param {object} [options={}]
+   * @property {string} options.search
+   * @property {string} options.search_fields - see OpenSearch Simple Query String
+   *                                        Query field argument for more information
+   * @property {integer} [options.page=1]
+   * @property {integer} [options.per_page=20]
```
Comment (wanglam) [inline src/plugins/workspace/public/workspace_client.ts:208]: The  type define in `WorkspaceFindOptions` is `perPage`, which one is more accurate?

File: src/plugins/workspace/public/workspace_client.ts [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    return [WORKSPACES_API_BASE_URL, join(...path)].filter((item) => item).join('/');
+  }
+
+  private async updateWorkspaceList(): Promise<void> {
+    const result = await this.list({
+      perPage: 999,
+    });
+
+    if (result?.success) {
+      const resultWithWritePermission = await this.list({
```
Comment (wanglam) [inline src/plugins/workspace/public/workspace_client.ts:104]: We don't have permission concepts in this PR, do we need to refactor here?

File: src/plugins/workspace/public/workspace_client.ts [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+      }
+
+      return {
+        success: false,
+        error: 'Unknown error',
+      };
+    }
+  };
+
+  private getPath(path: Array<string | undefined>): string {
```
Comment (wanglam) [inline src/plugins/workspace/public/workspace_client.ts:94]: How about `getPath(...path: Array<string | undefined>)`?

File: src/plugins/workspace/public/plugin.test.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      ]) as any;
+    });
+
+    const workspacePlugin = new WorkspacePlugin();
+    await workspacePlugin.setup(setupMock);
+    expect(setupMock.application.register).toBeCalledTimes(1);
+    expect(WorkspaceClientMock).toBeCalledTimes(1);
+    expect(workspaceClientMock.enterWorkspace).toBeCalledWith('workspaceId');
+    expect(setupMock.getStartServices).toBeCalledTimes(1);
+    await new Promise((resolve: any) => setTimeout(resolve, 1000));
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/plugin.test.ts:60]: I tried but there is an error saying container should be a dom, but got undefined. `waitFor` seems to be a react related test command.

File: src/plugins/workspace/public/plugin.test.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+          },
+        },
+        {},
+        {},
+      ]) as any;
+    });
+
+    const workspacePlugin = new WorkspacePlugin();
+    await workspacePlugin.setup(setupMock);
+    await new Promise((resolve: any) => setTimeout(resolve, 0));
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/plugin.test.ts:104]: No, good catch and I will remove this one.

File: src/plugins/workspace/public/plugin.test.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    currentAppIdSubscriber?.next(WORKSPACE_FATAL_ERROR_APP_ID);
+    await new Promise((resolve: any) => setTimeout(resolve, 0));
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/plugin.test.ts:106]: Sure, delete these two delay.

File: src/plugins/workspace/public/workspace_client.ts (hunk 1 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/workspace_client.ts:208]: perPage is correct, fixed.

File: src/plugins/workspace/public/workspace_client.ts (hunk 2 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/workspace_client.ts]: You are right, refractored.

File: src/plugins/workspace/public/workspace_client.ts (hunk 3 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/workspace_client.ts]: Agree, updated.

File: src/core/public/utils/workspace.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+export const formatUrlWithWorkspaceId = (
+  url: string,
+  workspaceId: string,
+  basePath?: IBasePath
+) => {
+  const newUrl = new URL(url, window.location.href);
+  /**
+   * Patch workspace id into path
+   */
+  newUrl.pathname = basePath?.remove(newUrl.pathname) || '';
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/core/public/utils/workspace.ts]: Nope, updated.

File: src/plugins/workspace/server/plugin.ts [hunk 8]
```diff
@@ -17,6 +17,23 @@ export class WorkspacePlugin implements Plugin<{}, {}> {
   private readonly logger: Logger;
   private client?: IWorkspaceDBImpl;
 
+  private proxyWorkspaceTrafficToRealHandler(setupDeps: CoreSetup) {
+    /**
+     * Proxy all {basePath}/w/{workspaceId}{osdPath*} paths to {basePath}{osdPath*}
+     */
+    setupDeps.http.registerOnPreRouting(async (request, response, toolkit) => {
+      const regexp = /\/w\/([^\/]*)/;
+      const matchedResult = request.url.pathname.match(regexp);
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/plugin.ts]: Sure, done for that.

File: src/plugins/workspace/server/plugin.ts [hunk 9]
```diff
@@ -17,6 +17,23 @@ export class WorkspacePlugin implements Plugin<{}, {}> {
...
+
+      if (matchedResult) {
+        const requestUrl = new URL(request.url.toString());
+        requestUrl.pathname = requestUrl.pathname.replace(regexp, '');
+        return toolkit.rewriteUrl(requestUrl.toString());
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/server/plugin.ts:31]: Sure, done for that.

File: src/plugins/workspace/public/workspace_client.ts (hunk 1 above)
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/workspace_client.ts]: It should have been fixed as well.

File: src/core/public/utils/workspace.ts [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  if (workspaceId) {
+    newUrl.pathname = `${WORKSPACE_PATH_PREFIX}/${workspaceId}${newUrl.pathname}`;
+  } else {
+    newUrl.pathname = newUrl.pathname.replace(/^\/w\/([^\/]*)/, '');
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/core/public/utils/workspace.ts]: Sure, we can reuse the clear workspace id function in proxy method.

File: src/plugins/workspace/public/workspace_client.ts [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+    });
+  };
+
+  /**
+   * Fetches a single workspace
+   *
+   * @param {string} id
+   * @returns The workspace for the given id.
+   */
+  public async get(id: string): Promise<IResponse<WorkspaceAttribute>> {
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/workspace_client.ts:222]: No, updated.

File: src/core/utils/workspace.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+   */
+  newUrl.pathname = basePath.remove(newUrl.pathname);
+
+  if (workspaceId) {
+    newUrl.pathname = `${WORKSPACE_PATH_PREFIX}/${workspaceId}${newUrl.pathname}`;
+  } else {
+    newUrl.pathname = cleanWorkspaceId(newUrl.pathname);
+  }
+
+  newUrl.pathname =
```
Comment (wanglam) [inline src/core/utils/workspace.ts:37]: Since `basePath` is not optional here, how about remove this `||`?

**Review by ruanyl — APPROVED** [review]

File: src/core/utils/workspace.ts (hunk 12 above)
Comment (SuZhou-Joe) [reply src/core/utils/workspace.ts]: Sure, done.

**Review by wanglam — APPROVED** [review]

File: src/plugins/workspace/public/components/workspace_fatal_error/workspace_fatal_error.tsx [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+import { IBasePath } from 'opensearch-dashboards/public';
+import { useOpenSearchDashboards } from '../../../../opensearch_dashboards_react/public';
+import { formatUrlWithWorkspaceId } from '../../../../../core/public/utils';
+
+export function WorkspaceFatalError(props: { error?: string }) {
+  const {
+    services: { application, http },
+  } = useOpenSearchDashboards();
+  const goBackToHome = () => {
+    window.location.href = formatUrlWithWorkspaceId(
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/plugins/workspace/public/components/workspace_fatal_error/workspace_fatal_error.tsx:25]: Actually I do not know if we should do that as I think `getUrlForApp` is a method to handle path without knowing `basePath`, however our workspace id is inside basePath.

[the remark this answers is not in the document]
Comment (ruanyl) [reply src/plugins/workspace/public/components/workspace_fatal_error/workspace_fatal_error.tsx:25]: I see, that makes sense

**Pull request merged**